<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Tommaso Rigon">

<title>Shrinkage and variable selection</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="un_C_files/libs/clipboard/clipboard.min.js"></script>
<script src="un_C_files/libs/quarto-html/quarto.js"></script>
<script src="un_C_files/libs/quarto-html/popper.min.js"></script>
<script src="un_C_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="un_C_files/libs/quarto-html/anchor.min.js"></script>
<link href="un_C_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="un_C_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="un_C_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="un_C_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="un_C_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<style>html{ scroll-behavior: smooth; }</style>

  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@0.15.1/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.15.1/dist/katex.min.css">

</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#homepage" id="toc-homepage" class="nav-link active" data-scroll-target="#homepage">Homepage</a></li>
  <li><a href="#the-prostate-dataset" id="toc-the-prostate-dataset" class="nav-link" data-scroll-target="#the-prostate-dataset">The <code>prostate</code> dataset</a></li>
  <li><a href="#a-glimpse-of-the-prostate-dataset" id="toc-a-glimpse-of-the-prostate-dataset" class="nav-link" data-scroll-target="#a-glimpse-of-the-prostate-dataset">A <code>glimpse</code> of the <code>prostate</code> dataset</a></li>
  <li><a href="#correlation-matrix-of-prostate" id="toc-correlation-matrix-of-prostate" class="nav-link" data-scroll-target="#correlation-matrix-of-prostate">Correlation matrix of <code>prostate</code></a></li>
  <li><a href="#the-variable-selection-problem" id="toc-the-variable-selection-problem" class="nav-link" data-scroll-target="#the-variable-selection-problem">The variable selection problem</a></li>
  <li><a href="#a-naïve-approach-abusing-p-values" id="toc-a-naïve-approach-abusing-p-values" class="nav-link" data-scroll-target="#a-naïve-approach-abusing-p-values">A naïve approach: (ab)using p-values</a></li>
  <li><a href="#a-predictive-perspective" id="toc-a-predictive-perspective" class="nav-link" data-scroll-target="#a-predictive-perspective">A predictive perspective</a></li>
  <li><a href="#best-subset-selection" id="toc-best-subset-selection" class="nav-link" data-scroll-target="#best-subset-selection">Best subset selection</a></li>
  <li><a href="#step-1.-and-2.-of-best-subset-selection" id="toc-step-1.-and-2.-of-best-subset-selection" class="nav-link" data-scroll-target="#step-1.-and-2.-of-best-subset-selection">Step 1. and 2. of best subset selection</a></li>
  <li><a href="#step-3.-of-best-subset-selection-via-cross-validation" id="toc-step-3.-of-best-subset-selection-via-cross-validation" class="nav-link" data-scroll-target="#step-3.-of-best-subset-selection-via-cross-validation">Step 3. of best subset selection via cross-validation</a></li>
  <li><a href="#comments-and-computational-considerations" id="toc-comments-and-computational-considerations" class="nav-link" data-scroll-target="#comments-and-computational-considerations">Comments and computational considerations</a></li>
  <li><a href="#forward-and-backward-regression" id="toc-forward-and-backward-regression" class="nav-link" data-scroll-target="#forward-and-backward-regression">Forward and backward regression</a></li>
  <li><a href="#forward-and-backward-regression-1" id="toc-forward-and-backward-regression-1" class="nav-link" data-scroll-target="#forward-and-backward-regression-1">Forward and backward regression</a></li>
  <li><a href="#comments" id="toc-comments" class="nav-link" data-scroll-target="#comments">Comments</a></li>
  <li><a href="#principal-components" id="toc-principal-components" class="nav-link" data-scroll-target="#principal-components">Principal components</a>
  <ul class="collapse">
  <li><a href="#principal-component-analysis" id="toc-principal-component-analysis" class="nav-link" data-scroll-target="#principal-component-analysis">Principal component analysis</a></li>
  <li><a href="#principal-components-regression" id="toc-principal-components-regression" class="nav-link" data-scroll-target="#principal-components-regression">Principal components regression</a></li>
  <li><a href="#pcr" id="toc-pcr" class="nav-link" data-scroll-target="#pcr">PCR</a></li>
  <li><a href="#a-summary-and-next-directions" id="toc-a-summary-and-next-directions" class="nav-link" data-scroll-target="#a-summary-and-next-directions">A summary and next directions</a></li>
  </ul></li>
  <li><a href="#ridge-regression" id="toc-ridge-regression" class="nav-link" data-scroll-target="#ridge-regression">Ridge regression</a>
  <ul class="collapse">
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="un_C_slides.html"><i class="bi bi-file-slides"></i>RevealJS</a></li></ul></div></nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Shrinkage and variable selection</h1>
<p class="subtitle lead">Data Mining - CdL CLAMSES</p>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliation</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author"><span class="orange">Tommaso Rigon</span> </p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            <em>Università degli Studi di Milano-Bicocca</em>
          </p>
      </div>
    </div>

<div class="quarto-title-meta">

      
  
    
  </div>
  

</header>

<section id="homepage" class="level2">
<h2 class="anchored" data-anchor-id="homepage"><a href="../index.html">Homepage</a></h2>
<div class="columns">
<div class="column" style="width:25%;">
<p><img src="img/lasso.png" class="img-fluid"></p>
</div><div class="column" style="width:75%;">
<ul>
<li><p>In this unit we will cover the following <span class="orange">topics</span>:</p>
<ul>
<li>Best subset regression</li>
<li>Principal component regression</li>
<li>Ridge regression</li>
<li>Lasso, LARS, elastic-net</li>
</ul></li>
<li><p>The common thread among these topics is the so-called <span class="blue">variable selection</span> problem.</p></li>
<li><p>The issue we face is the presence of several <span class="blue">potentially irrelevant</span> variables among the covariates.</p></li>
<li><p>In the <span class="orange">extreme case</span> <span class="math inline">p &gt; n</span>, is there even hope that we can fit a useful model?</p></li>
</ul>
</div>
</div>
</section>
<section id="the-prostate-dataset" class="level2">
<h2 class="anchored" data-anchor-id="the-prostate-dataset">The <code>prostate</code> dataset</h2>
<div class="incremental">
<ul class="incremental">
<li><p>The <code>prostate</code> cancer data investigates the relationship between the prostate-specific <span class="orange">antigen</span> and a number of clinical measures, in men about to receive a prostatectomy.</p></li>
<li><p>We want to <span class="blue">predict</span> the logarithm of a <span class="orange">prostate-specific antigen</span> (<code>lpsa</code>) as a function of:</p>
<ul class="incremental">
<li>logarithm of the cancer volume (<code>lcavol</code>);</li>
<li>logarithm of the prostate weight (<code>lweight</code>);</li>
<li>age each man (<code>age</code>);</li>
<li>logarithm of the benign prostatic hyperplasia amount (<code>lbph</code>);</li>
<li>seminal vesicle invasion (<code>svi</code>), a binary variable;</li>
<li>logarithm of the capsular penetration (<code>lcp</code>);</li>
<li>Gleason score (<code>gleason</code>), an ordered categorical variable;</li>
<li>Percentage of Gleason scores <span class="math inline">4</span> and <span class="math inline">5</span> (<code>pgg45</code>).</li>
</ul></li>
</ul>
</div>
</section>
<section id="a-glimpse-of-the-prostate-dataset" class="level2">
<h2 class="anchored" data-anchor-id="a-glimpse-of-the-prostate-dataset">A <code>glimpse</code> of the <code>prostate</code> dataset</h2>
<ul>
<li>The dataset is <a href="https://hastie.su.domains/ElemStatLearn/datasets/prostate.data">available online</a>. A description is given in <span class="blue">Section 3.2.1</span> of HTF (2009).</li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 97
Columns: 10
$ lcavol  &lt;dbl&gt; -0.5798185, -0.9942523, -0.5108256, -1.2039728, 0.7514161, -1.…
$ lweight &lt;dbl&gt; 2.769459, 3.319626, 2.691243, 3.282789, 3.432373, 3.228826, 3.…
$ age     &lt;int&gt; 50, 58, 74, 58, 62, 50, 64, 58, 47, 63, 65, 63, 63, 67, 57, 66…
$ lbph    &lt;dbl&gt; -1.3862944, -1.3862944, -1.3862944, -1.3862944, -1.3862944, -1…
$ svi     &lt;int&gt; 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,…
$ lcp     &lt;dbl&gt; -1.3862944, -1.3862944, -1.3862944, -1.3862944, -1.3862944, -1…
$ gleason &lt;int&gt; 6, 6, 7, 6, 6, 6, 6, 6, 6, 6, 6, 6, 7, 7, 7, 6, 7, 6, 6, 6, 6,…
$ pgg45   &lt;int&gt; 0, 0, 20, 0, 0, 0, 0, 0, 0, 0, 0, 0, 30, 5, 5, 0, 30, 0, 0, 0,…
$ lpsa    &lt;dbl&gt; -0.4307829, -0.1625189, -0.1625189, -0.1625189, 0.3715636, 0.7…
$ train   &lt;lgl&gt; TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, FALSE, TRUE, FALSE, FALSE,…</code></pre>
</div>
</div>
<ul>
<li><p>There are in total <span class="math inline">8</span> <span class="orange">variables</span> that can be used to predict the antigen <code>lpsa</code>.</p></li>
<li><p>The variable <code>train</code> splits the data into a training and test set, as in the textbook.</p></li>
<li><p>There are <span class="math inline">n = 67</span> observations in the <span class="orange">training</span> set and <span class="math inline">30</span> in the <span class="blue">test</span> set.</p></li>
</ul>
</section>
<section id="correlation-matrix-of-prostate" class="level2">
<h2 class="anchored" data-anchor-id="correlation-matrix-of-prostate">Correlation matrix of <code>prostate</code></h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="un_C_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="2250"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="the-variable-selection-problem" class="level2">
<h2 class="anchored" data-anchor-id="the-variable-selection-problem">The variable selection problem</h2>
<div class="incremental">
<ul class="incremental">
<li><p>We consider a <span class="orange">linear model</span> in which the response variable <span class="math inline">Y_i</span> (<code>lpsa</code>) is related to the covariates through the function<span class="math display">
  f(\bm{x}_i; \beta_0, \beta) = \beta_0+ \beta_1 x_{i1} + \cdots + \beta_p x_{ip} =\beta_0 + \bm{x}_i^T\beta.
  </span> In this unit the <span class="blue">intercept</span> <span class="math inline">\beta_0</span> will play a special role, therefore we use a slightly different notation compared to <a href="unit_A1.html">Unit A.1</a>.</p></li>
<li><p>If we use all the <span class="math inline">p = 8</span> available covariates, the estimated <span class="math inline">f(\bm{x}; \hat{\beta_0}, \hat{\beta})</span> might have a <span class="orange">high variance</span>, without important gain in term of bias, i.e.&nbsp;a <span class="blue">large mean squared error</span>.</p></li>
<li><p>Indeed, some variables are likely to be <span class="blue">irrelevant</span>:</p>
<ul class="incremental">
<li>they might be <span class="orange">correlated</span> with other covariates and therefore <span class="orange">redundant</span>;</li>
<li>they could be uncorrelated with the response <code>lpsa</code>.</li>
</ul></li>
<li><p>Hence, we are looking for a <span class="orange">simpler model</span> having hopefully a lower mean squared error.</p></li>
</ul>
</div>
</section>
<section id="a-naïve-approach-abusing-p-values" class="level2">
<h2 class="anchored" data-anchor-id="a-naïve-approach-abusing-p-values">A naïve approach: (ab)using p-values</h2>
<div style="font-size: 75%;">
<div class="cell">
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 13%">
<col style="width: 16%">
<col style="width: 9%">
<col style="width: 10%">
<col style="width: 8%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 8%">
<col style="width: 10%">
<col style="width: 8%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">(Intercept)</th>
<th style="text-align: right;">lcavol</th>
<th style="text-align: right;">lweight</th>
<th style="text-align: right;">age</th>
<th style="text-align: right;">lbph</th>
<th style="text-align: right;">svi</th>
<th style="text-align: right;">lcp</th>
<th style="text-align: right;">gleason</th>
<th style="text-align: right;">pgg45</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">estimate</td>
<td style="text-align: right;">2.46</td>
<td style="text-align: right;">0.68</td>
<td style="text-align: right;">0.26</td>
<td style="text-align: right;">-0.14</td>
<td style="text-align: right;">0.21</td>
<td style="text-align: right;">0.31</td>
<td style="text-align: right;">-0.29</td>
<td style="text-align: right;">-0.02</td>
<td style="text-align: right;">0.27</td>
</tr>
<tr class="even">
<td style="text-align: left;">std.error</td>
<td style="text-align: right;">0.09</td>
<td style="text-align: right;">0.13</td>
<td style="text-align: right;">0.10</td>
<td style="text-align: right;">0.10</td>
<td style="text-align: right;">0.10</td>
<td style="text-align: right;">0.12</td>
<td style="text-align: right;">0.15</td>
<td style="text-align: right;">0.15</td>
<td style="text-align: right;">0.15</td>
</tr>
<tr class="odd">
<td style="text-align: left;">statistic</td>
<td style="text-align: right;">27.60</td>
<td style="text-align: right;">5.37</td>
<td style="text-align: right;">2.75</td>
<td style="text-align: right;">-1.40</td>
<td style="text-align: right;">2.06</td>
<td style="text-align: right;">2.47</td>
<td style="text-align: right;">-1.87</td>
<td style="text-align: right;">-0.15</td>
<td style="text-align: right;">1.74</td>
</tr>
<tr class="even">
<td style="text-align: left;">p.value</td>
<td style="text-align: right;">0.00</td>
<td style="text-align: right;">0.00</td>
<td style="text-align: right;">0.01</td>
<td style="text-align: right;">0.17</td>
<td style="text-align: right;">0.04</td>
<td style="text-align: right;">0.02</td>
<td style="text-align: right;">0.07</td>
<td style="text-align: right;">0.88</td>
<td style="text-align: right;">0.09</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="incremental">
<ul class="incremental">
<li><p>It is a common practice to use the <span class="orange">p-values</span>, e.g.&nbsp;those obtained through the <code>summary</code> function, to perform <span class="blue">model selection</span> in a stepwise fashion.</p></li>
<li><p>A typical procedure is to omit “non significant” coefficients, refit the model, and repeat this scheme until we obtain only “significant” coefficients.</p></li>
<li><p>This is <span class="orange">not the best idea</span>, at least without careful thinking and appropriate multiplicity corrections.</p></li>
<li><p>In the first place, the p-values in the above table are meant to be used in the context of a single hypothesis testing problem, not to make several iterative choices.</p></li>
<li><p>Such an iterative usage of p-values is formally <span class="orange">incorrect</span> because it leads to the well-known <span class="blue">multiple testing problem</span>, which would require suitable corrections.</p></li>
</ul>
</div>
</section>
<section id="a-predictive-perspective" class="level2">
<h2 class="anchored" data-anchor-id="a-predictive-perspective">A predictive perspective</h2>
<ul>
<li>“<em>All models are approximations. Essentially, all models are wrong, but some are useful</em>.” George E.P. Box</li>
</ul>
<div class="incremental">
<ul class="incremental">
<li><p>There is another important reason for avoiding p-values. If the <span class="blue">focus</span> is just on <span class="blue">prediction</span>, we do not necessarily care about selecting the “true” set of parameters.</p></li>
<li><p>In many data mining problems, the focus is on the <span class="orange">minimization</span> of the <span class="orange">prediction errors</span>.</p></li>
<li><p>Hence, often times we may <span class="blue">accept some bias</span> (i.e.&nbsp;we use a “wrong” but useful model), if this leads to a <span class="orange">reduction in variance</span>.</p></li>
<li><p>Besides, in certain cases it does not even make much sense to speak about the “true parameters”.</p></li>
<li><p>For example, what if the true <span class="math inline">f(\bm{x})</span> were not linear? In this context, a <span class="blue">linear model</span> is simply an approximation of the unknown <span class="math inline">f(\bm{x})</span> and hypothesis testing procedures are ill-posed.</p></li>
</ul>
</div>
</section>
<section id="best-subset-selection" class="level2">
<h2 class="anchored" data-anchor-id="best-subset-selection">Best subset selection</h2>
<ul>
<li><p>Instead of choosing the optimal model using p-values, a more principled approach is based on the tools of <a href="un_B.html">Unit B</a>.</p></li>
<li><p>Ideally, we could perform an <span class="orange">exhaustive search</span> considering all the <span class="math inline">2^p</span> possible models and then picking the one having the best out-of-sample predictive performance.</p></li>
</ul>
<div class="callout callout-style-simple callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Best subset procedure
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li><p>Let <span class="math inline">\mathcal{M}_0</span> be the <span class="blue">null model</span>, which contains no predictors (just <span class="math inline">\hat{\beta}_0 = \bar{y}</span>).</p></li>
<li><p>For <span class="math inline">k =1,\dots,p</span>, do:</p>
<ol type="i">
<li><p>Estimate <span class="orange">all</span> the <span class="math inline">\binom{p}{k}</span> models that contain exactly <span class="math inline">k</span> covariates;</p></li>
<li><p>Identify the “best” model <span class="math inline">\mathcal{M}_k</span> with <span class="math inline">k</span> covariates having the smallest <span class="math inline">\text{MSE}_{k, \text{train}}</span>.</p></li>
</ol></li>
</ol>
</div>
</div>
<ul>
<li>By construction, a model with more variables <span class="math inline">\text{MSE}_{k + 1, \text{train}} \le \text{MSE}_{k, \text{train}}</span> has lower <span class="orange">training</span> error. Hence, the optimal subset size <span class="math inline">k</span> is chosen e.g.&nbsp;via <span class="blue">cross-validation</span>.</li>
</ul>
</section>
<section id="step-1.-and-2.-of-best-subset-selection" class="level2">
<h2 class="anchored" data-anchor-id="step-1.-and-2.-of-best-subset-selection">Step 1. and 2. of best subset selection</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="un_C_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="1500"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="step-3.-of-best-subset-selection-via-cross-validation" class="level2">
<h2 class="anchored" data-anchor-id="step-3.-of-best-subset-selection-via-cross-validation">Step 3. of best subset selection via cross-validation</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="un_C_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="1500"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="comments-and-computational-considerations" class="level2">
<h2 class="anchored" data-anchor-id="comments-and-computational-considerations">Comments and computational considerations</h2>
<ul>
<li>asd</li>
</ul>
</section>
<section id="forward-and-backward-regression" class="level2">
<h2 class="anchored" data-anchor-id="forward-and-backward-regression">Forward and backward regression</h2>
</section>
<section id="forward-and-backward-regression-1" class="level2">
<h2 class="anchored" data-anchor-id="forward-and-backward-regression-1">Forward and backward regression</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="un_C_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="1500"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="comments" class="level2">
<h2 class="anchored" data-anchor-id="comments">Comments</h2>
</section>
<section id="principal-components" class="level1">
<h1>Principal components</h1>
<section id="principal-component-analysis" class="level2">
<h2 class="anchored" data-anchor-id="principal-component-analysis">Principal component analysis</h2>
</section>
<section id="principal-components-regression" class="level2">
<h2 class="anchored" data-anchor-id="principal-components-regression">Principal components regression</h2>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'pls'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:stats':

    loadings</code></pre>
</div>
</div>
</section>
<section id="pcr" class="level2">
<h2 class="anchored" data-anchor-id="pcr">PCR</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="un_C_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="1500"></p>
</figure>
</div>
</div>
</div>
<!-- ## Summary of the estimated coefficients -->
<!-- ::: {style="font-size: 70%;"} -->
<!-- ```{r} -->
<!-- library(DT) -->
<!-- tab <- data.frame(OLS = rep(0, p), best_subset = rep(0, p), PCR = rep(0, p)) -->
<!-- rownames(tab) <- colnames(sum_best$which) -->
<!-- tab$OLS <- coef(lm(lpsa ~ ., data = prostate_train)) -->
<!-- tab$best_subset <- c(coef(lm(lpsa ~ lcavol + lweight, data = prostate_train)), rep(0, 6)) -->
<!-- # Principal components regression (PCR) -->
<!-- fit_pcr <- pcr(lpsa ~ ., data = prostate_train, center = TRUE, scale = FALSE) -->
<!-- beta <- c(coef(fit_pcr, 3)) -->
<!-- beta <- c(mean(prostate_train$lpsa) - colMeans(X[, -1]) %*% beta, beta) -->
<!-- tab$PCR <- beta -->

<!-- datatable(tab, colnames = c("OLS", "Best subset", "PCR"), options = list( -->
<!--   pageLength = 9, -->
<!--   dom = "t")) %>% -->
<!--   formatRound(columns = 1:3, digits = 3) %>% -->
<!--   formatStyle( -->
<!--     columns = 0, fontWeight = "bold" -->
<!--   ) %>% -->
<!--   formatStyle( -->
<!--     columns = 1:3, -->
<!--     backgroundColor = styleInterval(0, c("#FED8B1", "#DBE9FA")) -->
<!--   ) %>% -->
<!--   formatStyle( -->
<!--     columns = 1:3, -->
<!--     backgroundColor = styleEqual(0, c("white")) -->
<!--   ) -->
<!-- ``` -->
<!-- ::: -->
</section>
<section id="a-summary-and-next-directions" class="level2">
<h2 class="anchored" data-anchor-id="a-summary-and-next-directions">A summary and next directions</h2>
<table class="table">
<colgroup>
<col style="width: 17%">
<col style="width: 47%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><span class="orange">Shrinkage</span></th>
<th><span class="orange">Variable selection</span></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><span class="blue">Discrete</span></td>
<td>Principal component regression</td>
<td>Best subset selection</td>
</tr>
<tr class="even">
<td><span class="blue">Continuous</span></td>
<td>Ridge regression</td>
<td>Lasso</td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="ridge-regression" class="level1">
<h1>Ridge regression</h1>
<!-- ## Ridge regression -->
<!-- ```{r} -->
<!-- df_ridge <- function(lambda, X) { -->
<!--   X_tilde <- scale(X, TRUE, FALSE) -->
<!--   d2 <- eigen(crossprod(X_tilde))$values -->
<!--   sum(d2 / (d2 + lambda)) -->
<!-- } -->
<!-- df_ridge <- Vectorize(df_ridge, vectorize.args = "lambda") -->
<!-- ``` -->
<!-- ```{r} -->
<!-- library(glmnet) -->
<!-- my_ridge <- function(X, y, lambda){ -->
<!--   n <- nrow(X) -->
<!--   p <- ncol(X) -->
<!--   y_mean <- mean(y) -->
<!--   y <- y - y_mean -->
<!--   X_mean <- colMeans(X) -->
<!--   X <- X - rep(1,n) %*% t(X_mean) -->
<!--   X_scale <- sqrt( diag( (1/n) * crossprod(X) ) ) -->
<!--   X <- X %*% diag( 1 / X_scale ) -->
<!--   beta_scaled <- solve(crossprod(X) + lambda*diag(rep(1,p)), t(X) %*% y)  -->
<!--   beta <- diag( 1 / X_scale ) %*% beta_scaled -->
<!--   beta0 <- y_mean - X_mean %*% beta -->
<!--   return(c(beta0, beta)) -->
<!-- } -->
<!-- l = 1 -->
<!-- my_ridge(X,y,lambda = l) -->
<!-- coef(glmnet(X, y, alpha=0, lambda = l/n, thresh = 1e-20)) -->
<!-- y_std <- scale(y, center=TRUE, scale=sd(y)*sqrt((n-1)/n) )[,] -->
<!-- my_ridge(X,y_std,lambda = l) -->
<!-- coef(glmnet(X, y_std, alpha=0, lambda = l/n, thresh = 1e-20)) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- #  -->
<!-- #  -->
<!-- # lambda <- 100 -->
<!-- # XX <- X[,-1] #scale(X[, -1], TRUE, scale = apply(X[, -1], 2, function(x)  sqrt(mean(x^2) - mean(x)^2))) -->
<!-- # yy <- y #(y - mean(y)) / sqrt(mean(y^2) - mean(y)^2) -->
<!-- #  -->
<!-- # cv_ridge_fit <- cv.glmnet(XX, yy, family = "gaussian", standardize = FALSE, lambda = exp(seq(-10, 12, length = 500)), -->
<!-- #                     alpha = 0, thresh = 1e-16) -->
<!-- # plot(cv_ridge_fit) -->
<!-- #  -->
<!-- # c(solve((crossprod(XX) + lambda * diag(p-1)), crossprod(XX, yy))) -->
<!-- # c(coef(fit_ridge)[-1, ]) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- # plot(log(cv_ridge_fit$lambda), cv_ridge_fit$cvm, type = "l") -->
<!-- # plot(1 + df_ridge(nrow(X[, -1]) * cv_ridge_fit$lambda, X[, -1]),  -->
<!-- #      cv_ridge_fit$cvm, type = "b", xlab = "Model complexity (p)", ylab = "MSE") -->
<!-- # lines(1 + df_ridge(nrow(X[, -1]) * cv_ridge_fit$lambda, X[, -1]),  -->
<!-- #       cv_ridge_fit$cvup, type = "b", xlab = "Model complexity (p)", ylab = "MSE", lty = "dashed", col = "red") -->
<!-- #  -->
<!-- # 1 + df_ridge(nrow(X[, -1]) * cv_ridge_fit$lambda.min, X[, -1]) -->
<!-- # 1 + df_ridge(nrow(X[, -1]) * cv_ridge_fit$lambda.1se, X[, -1]) -->
<!-- #  -->
<!-- # ridge_fit <- cv_ridge_fit$glmnet.fit -->
<!-- # coef(ridge_fit) -->
<!-- # plot(ridge_fit, , label = TRUE) -->
<!-- ``` -->
<!-- ## The wrong way of doing cross-validation -->
<!-- ::: incremental -->
<!-- -   Consider a regression problem with a [large number of -->
<!--     predictors]{.blue}, as may arise, for example, in genomic or -->
<!--     proteomic applications. -->
<!-- -   A typical strategy for analysis might be as follows: -->
<!--     1.  Screen the predictors: find a subset of "good" predictors that -->
<!--         show fairly strong (univariate) correlation with the class -->
<!--         labels; -->
<!--     2.  Using just this subset of predictors, build a regression model; -->
<!--     3.  Use cross-validation to estimate the unknown tuning parameters -->
<!--         (i.e. degree of polynomials) and to estimate the prediction -->
<!--         error of the final model. -->
<!-- -   Is this a correct application of cross-validation? -->
<!-- -   If your reaction was "[this is absolutely wrong!]{.orange}", it -->
<!--     means you correctly understood the principles of cross-validation. -->
<!-- -   If you though this was an ok-ish idea, please read [Section -->
<!--     7.10.2]{.blue} of HTF (2009). -->
<!-- ::: -->
<!-- # Lasso, LARS, and elastic-net -->
<!-- ## Lasso -->
<!-- ::: columns -->
<!-- ::: {.column width="25%"} -->
<!-- ![](img/lasso.png) -->
<!-- ::: -->
<!-- ::: {.column width="75%"} -->
<!-- -   asdasd -->
<!-- ::: -->
<!-- ::: -->
<!-- ## Lasso -->
<!-- ```{r} -->
<!-- library(lars) -->
<!-- lambda <- 100 -->
<!-- XX <- scale(X[, -1], TRUE, scale = apply(X[, -1], 2, function(x)  sqrt(mean(x^2) - mean(x)^2))) -->
<!-- yy <- y#(y - mean(y)) / sqrt(mean(y^2) - mean(y)^2) -->
<!-- cv_lars <- cv.lars(x = XX, y = yy, K = 10, type = "lasso", mode = "step") -->
<!-- fit_lars <- lars(x = XX, y = yy, type ="lasso", normalize = FALSE) -->
<!-- cv_lasso_fit <- cv.glmnet(XX, yy, standardize = FALSE,  -->
<!--                           family = "gaussian", alpha = 1, nfolds = 10, -->
<!--                           lambda = 1 / n * fit_lars$lambda, thresh = 1e-16) -->
<!-- plot(cv_lasso_fit) -->
<!-- lasso_fit <- cv_lasso_fit$glmnet.fit -->
<!-- lambda_sel <- 4 -->
<!-- round(coef(fit_lars)[lambda_sel, ], 5) -->
<!-- round(coef(lasso_fit, mode = "lambda")[-1, lambda_sel], 5) -->
<!-- ``` -->
<section id="references" class="level2">
<h2 class="anchored" data-anchor-id="references">References</h2>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



<script src="un_C_files/libs/quarto-html/zenscroll-min.js"></script>
</body></html>